src/include/server/mir/scene/text_input_hub.h

```diff
@@ -97,6 +99,47 @@ struct TextInputState
     std::optional<TextInputContentPurpose> content_purpose;
 };
 
+class CopyableWlArray
+{
+public:
+    explicit CopyableWlArray(wl_array* in_data)
+        : data_{copy(in_data)}
+    {
+    }
+
+    CopyableWlArray(CopyableWlArray const& other)
+        : CopyableWlArray{other.data_}
+    {
+    }
+
+    CopyableWlArray& operator=(CopyableWlArray other)
+    {
+        std::swap(data_, other.data_);
+        return *this;
+    }
+
+    ~CopyableWlArray()
+    {
+        wl_array_release(data_);
+    }
+
+    [[nodiscard]] wl_array* data() const
+    {
+        return data_;
+    }
+
+private:
+    static wl_array* copy(wl_array* in_data)
+    {
+        auto data = new wl_array();
+        wl_array_init(data);
+        wl_array_copy(data, in_data);
+        return data;
+    }
+
+    wl_array* data_;
+};
```

Comment:
This fails to delete `data_`, but the whole thing can be further simplfied...
```suggestion
class CopyableWlArray
{
public:
    explicit CopyableWlArray(wl_array const* in_data)
    {
        wl_array_init(&data_);
        wl_array_copy(&data_, const_cast<wl_array*>(in_data)); // wl_array_copy() doesn't mark source as const
    }

    CopyableWlArray(CopyableWlArray const& other)
        : CopyableWlArray{&other.data_}
    {
    }

    CopyableWlArray& operator=(CopyableWlArray other)
    {
        std::swap(data_, other.data_);
        return *this;
    }

    ~CopyableWlArray()
    {
        wl_array_release(&data_);
    }

    [[nodiscard]] wl_array* data() const
    {
        // Our generated `send_modifiers_map_event()` requires a non-const wl_array, so
        // this workaround may be the wrong thing
        return const_cast<wl_array*>(&data_);
    }

private:

    wl_array data_;
};
```